src/ir/effects.h

```diff
@@ -217,8 +217,32 @@ class EffectAnalyzer {
   // check if we break to anything external from ourselves
   bool hasExternalBreakTargets() const { return !breakTargets.empty(); }
 
-  // checks if these effects would invalidate another set (e.g., if we write, we
-  // invalidate someone that reads, they can't be moved past us)
+  // Checks if these effects would invalidate another set of effects (e.g., if
+  // we write, we invalidate someone that reads).
+  //
+  // This assumes the things whose effects we are comparing will both execute,
+  // at least if neither of them transfers control flow away. That is, we assume
+  // that there is no transfer of control flow *between* them: we are comparing
+  // things appear after each other, perhaps with some other code in the middle,
+  // but that code does not transfer control flow. For example, here it is ok to
+  // compare the effects of A and B, no matter what A and B are:
+  //
+  //   A
+  //   (local.set 0 (i32.const 0))
+  //   B
+  //
+  // But here it is not:
+  //
+  //   A
+  //   (br_if 0 (local.get 0)) ;; this may transfer control flow away
+  //   B
```

Comment:
In this case, if neither A nor B returns true for `transferscontrolFlow()`, because this `br_if` in between, we cannot reorder A and B, right? I understand we only compare adjacent expressions but then is this example in the comment appropriate?

Reply:
Correct, we also cannot reorder them in this case.

Are you saying a better example is something that we *can* reorder? I'm not sure that exists, now that I think of it. That is, for us to be able to reorder two things we do need them to be consecutive, in general. Unless they are trivial things like say a `local.get` and a `nop`. But maybe I misunderstood your question?

Reply:
This function assumes the two expressions are adjacent, and invalidates the expression when either of them contains a control flow, but this example in the comment seems to be where two expressions (A and B) are not adjacent. That was my question.

Reply:
I see now, thanks. This is an example of something that the function should not even be called with. I'll rephrase to make the clearer.